Addition of location share feature

tgxn opened the pull request:
@RocketChat/core 
Closes #3791

![image](https://cloud.githubusercontent.com/assets/293107/17423573/4778352a-5aee-11e6-9311-0b8b9105d5a7.png)
![image](https://cloud.githubusercontent.com/assets/293107/17423604/7d29bf54-5aee-11e6-897c-708abd9e6e7b.png)
![image](https://cloud.githubusercontent.com/assets/293107/17423583/590e9dba-5aee-11e6-92e2-80936a1a80a8.png)

Currently, it works by inserting a "location string" into your messagebox, letting you manually send it. I have added a hook to the messageRenderer in order to add the nice "static maps" when one of the "location stings" is detected, along with two settings to enable/disable it, and set your Google Maps API Key.

I am unsure of the process of this being supported in the Android/iOS app compilation (via Cordova).

rodrigok commented on packages/rocketchat-mapview/server/settings.coffee:
```diff
@@ -0,0 +1,3 @@
+Meteor.startup ->
+	RocketChat.settings.add 'MapView_Enabled', false, {type: 'boolean', group: 'Message', section: 'Google Maps', public: true, i18nLabel: 'MapView Enabled', i18nDescription: 'MapView_Enabled_Description'}
```
Can you use the pattern `MapView_Enabled` for i18nLabel like you are using for description? And can you add these keys in `en.i18n.json` file? If you do that our translators/contributors will be able to see theses strings to translate in lingohub.

tgxn replied:
I've updated it with keys, and added descriptions properly. I'm fairly sure I don't need to touch the other i18n files, as i only added my keys to the en.i18n.json file.

rodrigok replied:
Thanks @tgxn, we only need the keys in en files

engelgabriel commented:
This looks awesome. I just think we should have a dedicated property of the message objects for that, so it can be indexed.

rodrigok commented:
@tgxn that is awesome. Can you solve the conflicts?

rodrigok commented:
@engelgabriel we can solve that like usernames, parsing at the server side

tgxn commented:
Thanks for the feedback; I'm fairly new here and just starting to find my way around the codebase.

I've looked into what @engelgabriel  mentioned regarding message-properties, and I think this method of storing locations will be superior to the method i was previously using (hardcoded string as message.msg).

I've updated the Message model with an additional index "2dsphere" for the "location" property. This should allow some cool future enhancements on this feature. (Search, view all locations on map, etc)

I also added a pretty modal popup so location can be confirmed before sending.
![image](https://cloud.githubusercontent.com/assets/293107/17455457/b405f806-5bea-11e6-9470-4f449129086a.png)

rodrigok commented:
@tgxn looks great, @engelgabriel can we merge this?

rodrigok commented:
LGTM

tgxn commented:
I'm happy for it to be merged, however I'd like to resolve the issue in that it doesn't work as expected when running in the mobile cordova build. I'm not sure if permissions on the official app would need to be updated, or if it's a build paramerter inside the code somewheere?

rodrigok commented:
@tgxn what happens if we don't have this plugin? Can you access the api `navigator.geolocation`?

If you can't access you can test if the api is available and then enable this feature, then we can add the plugin to our mobile apps and this feature will be available when users get the updated app.

tgxn commented:
@rodrigok Yup, from what I can see, the location API seems to be unavailable in the Cordova environment.

I tried to load the Cordova repo+build tools but I'm having some trouble getting it to compile!
Perhaps one of you that already has it setup can test to see if enabling that plugin resolves the issue?

rodrigok commented:
@tgxn Can you test if the geolocation API is available, if not can you hide the button? Then we can merge your PR and we can add the plugin to cordova and test locally, and this feature will not break for not udpated apps too.

tgxn commented:
@rodrigok Yep, that sounds like a winner! Updated with the check and tested. Appears to work well and hide the button in the Cordova app.

May need to change the check when support is added to Cordova, but for now this works well. :)

rodrigok commented:
Awesome @tgxn, thank you so much.
